docker/reqmgr2ms/run.sh

```diff
@@ -1,4 +1,58 @@
 #!/bin/bash
+
+### Usage: Usage: run.sh [-e 'RSE Expression']
+### Usage:
+### Usage:   -e  <RSE Expression>   A Rucio RSE Expression to be passed to the service configuration
+### Usage:                          before service start
+### Usage:   -h <help>              Provides help to the current script
+### Usage:
+### Usage: Example: ./run.sh -e 'rse_type=DISK&country=US&tier=3&cms_type=real'
+### Usage: Example: ./run.sh
+### Usage:
+
+FULL_SCRIPT_PATH="$(realpath "${0}")"
+
+usage()
+{
+    echo -e $1
+    grep '^### Usage:' < $FULL_SCRIPT_PATH
+    exit 1
+}
+
+help()
+{
+    echo -e $1
+    grep '^### Usage:' < $FULL_SCRIPT_PATH
+    exit 0
+}
+
+### Searching for the mandatory and optional arguments:
+# export OPTIND=1
+while getopts ":e:h" opt; do
+    case ${opt} in
+        e)
+            rseExpr=${OPTARG}
+            # escaping all possible special characters in the RSE expression:
+            # NOTE: One should always start escaping the escaping chars first
+            rseExpr=${rseExpr//\\/\\\\}
```

Comment:
I'm afraid I do not follow this logic, but from your note, I assume you have extensively tested it.

Reply:
Hi Alan yes, I did test it. Because Rucio uses `\` as a set compliment operator it collides with it's shell purpose of an special character escape operator and because of that if start putting escape characters in front of the rest of the Rucios set operators `&` `|` first then you end up with random number of `\` operators in the string out of which you cannot tell the purpose of each other (was it an already existing Rucio set operator, was it a shell escape operator from the previous iteration). let me give you an example: 

If you do NOT escape the `\` operator first:
```
$ rseExpr='((cms_type=real&rse_type=DISK&(tier=2|tier=3))\country=US)'
$ rseExpr=${rseExpr//&/\\&}
$ rseExpr=${rseExpr//|/\\|}
$ rseExpr=${rseExpr//\\/\\\\}
]$ echo $rseExpr 
((cms_type=real\\&rse_type=DISK\\&(tier=2\\|tier=3))\\country=US)
```

If you do escape the `\` operator first: 
```
$ rseExpr='((cms_type=real&rse_type=DISK&(tier=2|tier=3))\country=US)'
$ rseExpr=${rseExpr//\\/\\\\}
$ rseExpr=${rseExpr//&/\\&}
$ rseExpr=${rseExpr//|/\\|}
$ echo $rseExpr 
((cms_type=real\&rse_type=DISK\&(tier=2\|tier=3))\\country=US)
```

Reply:
Thank you for these details, Todor. The escaping part is know to me, what I got lost on was the syntax of the normal slashes `/`, so this bash expression: `${rseExpr//&/\\&}`. I will google for that though ;)

Reply:
No need to google @amaltaro, `man bash` tells everything. Here is the regular shell variable expansion section from the manual with the relevant pattern for this case:
```
 ${parameter/pattern/string}
              Pattern substitution.  The pattern is expanded to produce a pattern just as in pathname expansion, Parameter is expanded and the longest match of pattern against its value
              is  replaced  with  string.   The match is performed using the rules described under Pattern Matching below.  If pattern begins with /, all matches of pattern are replaced
              with string.  Normally only the first match is replaced.  If pattern begins with #, it must match at the beginning of the expanded value of parameter.  If  pattern  begins
              with %, it must match at the end of the expanded value of parameter.  If string is null, matches of pattern are deleted and the / following pattern may be omitted.  If the
              nocasematch shell option is enabled, the match is performed without regard to the case of alphabetic characters.  If parameter is @ or *, the substitution operation is ap‐
              plied to each positional parameter in turn, and the expansion is the resultant list.  If parameter is an array variable subscripted with @ or *, the substitution operation
              is applied to each member of the array in turn, and the expansion is the resultant list.
```

just a little help on the peculiarities for the example at hand: starting with double `//` tells the expansion engine to substitute the pattern everywhere it have been found not just the first occurrence. Then is the `\\` escape of the escape operator itself so that  the shell may avoid expanding the operator in the `${rseExpr//\\/\\\\}` statement itself. then comes the long chain of escapes `\\\\` which need accumulate levels of escapes based on the consequent steps down the road until it reaches the final destination i.e.:
   * first level of escapes is again the expansion of the current expression `${rseExpr//\\/\\\\}`, just as above
   * second level of escapes is for the expansion that happens for passing the variable to `sed`
   * third level of escape is the internal `sed` expansion, which cannot be avoided 
   * and the forth `\` is not an escape it is just the symbol itself (the Rucio set operator)

Reply:
Sorry, you meant this expression : `${rseExpr//&/\\&}` ... but any way the logic and the sequence of `/` and escapes is exactly the same.